WIP: Add typing to configstore

just1602 opened the pull request:
Fixes #25

just1602 committed:
Ignore .mypy_cache/

just1602 committed:
Add basic typing stuff

just1602 committed:
Remove unsued imports

just1602 committed:
Merge remote-tracking branch 'origin/master' into typing

just1602 commented on configstore/store.py:
```diff
@@ -7,13 +19,13 @@ class SettingNotFoundException(Exception):
         self._backends += (backend,)
 
-    def get_setting(self, key, default=_no_default):
+    def get_setting(self, key: str, default=_no_default) -> str:
```
Ici, default devrait être un string. Sauf que si default peut être `Union[object, str]`, ça veut dire que le retour de la méthode peut être lui aussi `Union[object, str]`, mais ce n'est pas du tout ce qu'on veut.

J'aurais tendance à faire un _no_default avec une valeur en string, mais faut trouver un default qui ne se retrouverait pas dans un default. Tu en pense quoi @merwok ?

merwok replied:
`_no_default = '~~!!configstore-no-default!!~~'`

merwok commented on configstore/store.py:
```diff
@@ -1,3 +1,15 @@
+
+
+Backends = TypeVar('Backends', EnvVarBackend, DotenvBackend,
+                   AwsSsmBackend, DockerSecretBackend)
```
It’s not great to have to import all backends here.

Could we declare and interface or ABC or base type and use that?

just1602 replied:
I woudl really prefer to declare a base class or an interface. It was my first plan, but I wasn't sure if you'd like it and I saw the TypeVar in the documentation.

My toher questions is since we'll have a backend inteface. could we move all backend in a `backends.py` file with the `Backend` class ?

merwok replied:
Go for an ABC or whatever is recommended by typing / mypy.

I don’t see benefits in moving the concrete backend classes though.

Commit:
Use a `_no_default` str variable instead of object

This way, it's really easier to make typing work without ugly hack or
long Union typing

Commit:
Remove python 2.7 tests since typing isn't support

There is support for typing in python 2.7 but only in comments and we
don't want this

Commit:
Remove python 2.7 support from flit packaging config file

Commit:
Also drop the py35 support since not all type hint are supported

just1602 commented on .circleci/config.yml:
```diff
@@ -63,57 +61,6 @@ jobs:
       - store_test_results:
           path: ~/reports
 
-  test-py35:
```
Comme python 3.5 est juste en security status et que certaine type hint ne sont pas supporté dedans, mais le sont dans python 3.6 et 3.7, j'ai enlevé le support de 3.5.

Commit:
Tottaly remove python 3.5

Commit:
All backends inherint from the Bcakend base class

Commit:
Fix backends typing

Commit:
Add Backend abstract class test

merwok commented on .gitignore:
```diff
@@ -6,6 +6,8 @@ __pycache__
 /.tox/
 /venv/
+/.pyre/
+.pyre_configuration
```
Does this file live at the repo root?

just1602 replied:
Normally yes.

merwok commented on .circleci/config.yml:
```diff
@@ -63,57 +61,6 @@ jobs:
-          name: Test with Python 2.7
-          command: |
-            venv/bin/tox --sdistonly
-            venv/bin/tox -e py27 -- --junitxml=~/reports/tox/coverage.xml
```
There’s some config that can be removed from tox.ini too if you want to clean up.

merwok commented on configstore/backends/base.py:
```diff
@@ -0,0 +1,10 @@
+        pass
+
+    def get_setting(self, key: str) -> Optional[str]:
+        raise NotImplementedError
```
I would like this to be a real abstract base class, and also keep the duck typing (not require that classes inherit from the base class).  We could chat about this on Monday, the background and reasoning could be quite long for a review message!

(Duck typing / virtual subclasses can be handled by typing: https://mypy.readthedocs.io/en/stable/protocols.html#protocol-types )

merwok commented on configstore/backends/env_var.py:
```diff
@@ -1,7 +1,11 @@
+
+class EnvVarBackend(Backend):
+
+    def get_setting(self, key: str) -> Optional[str]:
```
Great job fixing the param names in get_setting methods!

merwok reviewed (changes requested):
This is great!  Let’s discuss how to type-check duck-typed backend classes.

Commit:
merge master
